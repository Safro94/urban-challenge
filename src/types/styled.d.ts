import 'styled-components';

interface IPalette {
	main: string;
	text: string;
}

interface IBreakpoints {
	mobile: string;
	tablet: string;
	laptop: string;
	desktop: string;
}

declare module 'styled-components' {
	export interface DefaultTheme {
		breakpoints: IBreakpoints;
		palette: {
			common: {
				background: string;
				white: string;
				black: string;
				grey: string;
			};
			primary: IPalette;
		};
	}
}
