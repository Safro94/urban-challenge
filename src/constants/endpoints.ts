export const SLOTS_ENDPOINT: string = '/slots.json';
export const WORKERS_ENDPOINT: string = '/workers.json';
export const AVAILABLE_WORKERS_ENDPOINT: string = '/available-workers.json';
