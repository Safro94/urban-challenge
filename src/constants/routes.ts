export const HOME = '/';
export const WORKERS = '/workers';
