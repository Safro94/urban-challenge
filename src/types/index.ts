export interface ISlot {
	id: number;
	localisedTime: string;
	price: string;
}

export interface IWorker {
	id: number;
	name: string;
	rating: string;
	isNew: boolean;
}

export interface ICartItem {
	id: string;
	slot: ISlot | null;
	worker: IWorker;
	price: number;
	amount: number;
}

export interface IAvailableWorker {
	slotId: number;
	availableWorkersIds: number[];
}

export interface IWorkerResponse {
	workers: IWorker[];
}

export interface ISlotsResponse {
	slots: ISlot[];
}

export interface IAvailableWorkerResponse {
	slot_id: number;
	availableWorker_ids: number[];
}

export interface IAvailableWorkersResponse {
	'available-workers': IAvailableWorkerResponse[];
}

export interface ISelectOption {
	label: string;
	value: string;
	[key: string]: unknown;
}

export enum RequestMethods {
	GET = 'get',
	POST = 'post',
	DELETE = 'delete',
}

export enum ButtonTypes {
	Primary = 'primary',
}
