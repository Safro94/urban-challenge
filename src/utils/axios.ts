import axios, { AxiosError, AxiosRequestConfig } from 'axios';
import https from 'https';

const opts: AxiosRequestConfig = {
	httpsAgent: new https.Agent({ rejectUnauthorized: false }),
	baseURL: process.env.REACT_APP_API_URL,
};

const interceptorError = (error: AxiosError) => Promise.reject(error.response);

const instance = axios.create(opts);

instance.interceptors.response.use(response => response, interceptorError);

export default instance;
