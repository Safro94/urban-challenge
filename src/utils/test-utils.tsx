import { FC, ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { BrowserRouter as Router } from 'react-router-dom';

import { theme } from '../theme';

const AllTheProviders: FC = ({ children }) => {
	return (
		<ThemeProvider theme={theme}>
			<Router>{children}</Router>
		</ThemeProvider>
	);
};

const customRender = (
	ui: ReactElement,
	options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
