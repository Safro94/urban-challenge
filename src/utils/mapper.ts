import { IAvailableWorker, IAvailableWorkerResponse } from '../types';

export const mapAvailableWorkers = (
	workers: IAvailableWorkerResponse[]
): IAvailableWorker[] => {
	return workers.map((worker: IAvailableWorkerResponse) => ({
		slotId: worker.slot_id,
		availableWorkersIds: worker.availableWorker_ids,
	}));
};
