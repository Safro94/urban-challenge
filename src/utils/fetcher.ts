import { AxiosResponse } from 'axios';
import {
	IAvailableWorkersResponse,
	ISlotsResponse,
	IWorkerResponse,
	RequestMethods,
} from '../types';

import axios from './axios';

interface IFetcher {
	method?: RequestMethods;
	url: string;
	data?: null | any;
	body?: null | any;
}

type Response = IWorkerResponse | ISlotsResponse | IAvailableWorkersResponse;

const fetcher = async ({
	method = RequestMethods.GET,
	url,
	body = null,
}: IFetcher): Promise<AxiosResponse<Response>> => {
	try {
		const res: AxiosResponse<Response> = await axios[method](url, body);
		return res;
	} catch (error) {
		throw error;
	}
};

export default fetcher;
