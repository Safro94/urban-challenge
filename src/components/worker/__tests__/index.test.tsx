import userEvent from '@testing-library/user-event';

import { render, screen } from '../../../utils/test-utils';

import Worker from '..';

import { HOME } from '../../../constants/routes';

import { IWorker } from '../../../types';

const mockHistoryPush = jest.fn();
jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'),
	useHistory: () => ({
		push: mockHistoryPush,
	}),
}));

jest.mock('uuid', () => ({
	...jest.requireActual('uuid'),
	v4: () => '1',
}));

const mockAddToCart = jest.fn();
jest.mock('../../../hooks/cart', () => {
	return {
		useCart: () => ({
			addToCart: mockAddToCart,
		}),
	};
});

const mockSlot = {
	id: 1,
	localisedTime: '10:00',
	price: '$20',
};
jest.mock('../../../hooks/application', () => {
	return {
		useApplication: () => ({
			selectedSlot: mockSlot,
		}),
	};
});

describe('Worker', () => {
	const index = 1;
	const getItemProps = () => {};
	const worker: IWorker = {
		id: 1,
		name: 'test',
		rating: '4.5',
		isNew: true,
	};

	it('should render correctly', () => {
		const { container } = render(
			<Worker item={worker} index={index} getItemProps={getItemProps} />
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	describe('Worker functionality', () => {
		it('should call addToCart when the button is clicked', () => {
			render(
				<Worker item={worker} index={index} getItemProps={getItemProps} />
			);

			const button = screen.getByRole('button', { name: /book/i });
			userEvent.click(button);

			expect(mockAddToCart).toHaveBeenCalledTimes(1);
			expect(mockAddToCart).toHaveBeenCalledWith({
				worker,
				id: '1',
				slot: mockSlot,
				price: 20,
				amount: 1,
			});

			// expect(mockHistoryPush).toHaveBeenCalledTimes(1);
			// expect(mockHistoryPush).toHaveBeenCalledWith(HOME);
		});
	});
});
