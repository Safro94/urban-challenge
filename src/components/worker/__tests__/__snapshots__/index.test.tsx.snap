// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Worker should render correctly 1`] = `
<li
  class="sc-gsTEea kELbDD"
>
  <div
    class="sc-dlfnuX kAgLnS"
  >
    <div
      class="sc-jSgvzq Pbzfh"
    >
      <img
        alt="Worker 1"
        class="sc-iBPTik cYyUHg"
        src="therapist.jpeg"
      />
    </div>
    <div
      class="sc-eCstlR irqCIl"
    >
      <div
        class="sc-gKseQn fcrZBH"
      >
        <span
          class="sc-fubCzh"
        >
          Name:
        </span>
        <span
          class="sc-fubCzh"
        >
          test
        </span>
      </div>
      <div
        class="sc-gKseQn fcrZBH"
      >
        <span
          class="sc-fubCzh"
        >
          Rating:
        </span>
        <span
          class="sc-fubCzh"
        >
          ⭐⭐⭐⭐
        </span>
      </div>
    </div>
    <div
      class="sc-hKgJUU jXFudM"
    >
      <button
        class="sc-bdfBQB LqdYh"
        type="button"
      >
        Book
      </button>
    </div>
  </div>
</li>
`;
